ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java

```diff
@@ -175,93 +179,95 @@ private int sizeExcludingBlobAndCrcSize() {
    * Construct the bufferToSend to serialize request metadata and other blob related information. The newly constructed
    * bufferToSend will not include the blob content as it's carried by the {@code blob} field in this class.
    */
-  private void prepareBuffer() {
-    bufferToSend = ByteBuffer.allocate(sizeExcludingBlobAndCrcSize());
+  @Override
+  protected void prepareBuffer() {
+    bufferToSend = PooledByteBufAllocator.DEFAULT.heapBuffer(sizeExcludingBlobAndCrcSize());
     writeHeader();
-    int crcStart = bufferToSend.position();
-    bufferToSend.put(blobId.toBytes());
+    int crcStart = bufferToSend.writerIndex();
+    bufferToSend.writeBytes(blobId.toBytes());
     BlobPropertiesSerDe.serializeBlobProperties(bufferToSend, properties);
-    bufferToSend.putInt(usermetadata.capacity());
-    bufferToSend.put(usermetadata);
-    bufferToSend.putShort((short) blobType.ordinal());
+    bufferToSend.writeInt(usermetadata.capacity());
+    bufferToSend.writeBytes(usermetadata);
+    bufferToSend.writeShort((short) blobType.ordinal());
     short keyLength = blobEncryptionKey == null ? 0 : (short) blobEncryptionKey.remaining();
-    bufferToSend.putShort(keyLength);
+    bufferToSend.writeShort(keyLength);
     if (keyLength > 0) {
-      bufferToSend.put(blobEncryptionKey);
+      bufferToSend.writeBytes(blobEncryptionKey);
     }
-    bufferToSend.putLong(blobSize);
-    crc.update(bufferToSend.array(), bufferToSend.arrayOffset() + crcStart, bufferToSend.position() - crcStart);
-    nioBuffersFromBlob = blob.nioBuffers();
-    for (ByteBuffer bb : nioBuffersFromBlob) {
+    bufferToSend.writeLong(blobSize);
+    crc.update(bufferToSend.array(), bufferToSend.arrayOffset() + crcStart, bufferToSend.writerIndex() - crcStart);
+    for (ByteBuffer bb : blob.nioBuffers()) {
       crc.update(bb);
       // change it back to 0 since we are going to write it to the channel later.
       bb.position(0);
     }
-    crcBuf.putLong(crc.getValue());
-    crcBuf.flip();
-    bufferToSend.flip();
+    crcByteBuf.writeLong(crc.getValue());
+
+    CompositeByteBuf compositeByteBuf = bufferToSend.alloc().compositeHeapBuffer(2 + blob.nioBufferCount());
+    compositeByteBuf.addComponent(true, bufferToSend);
+    if (blob instanceof CompositeByteBuf) {
+      Iterator<ByteBuf> iter = ((CompositeByteBuf) blob).iterator();
+      while (iter.hasNext()) {
+        compositeByteBuf.addComponent(true, iter.next());
+      }
+    } else {
+      compositeByteBuf.addComponent(true, blob);
+    }
+    compositeByteBuf.addComponent(true, crcByteBuf);
+    blob = null;
+    crcByteBuf = null;
+    bufferToSend = compositeByteBuf;
```

Comment:
The implementation overuses bufferToSend and caused some confusion when I looked at the code. 
How about just set 
bufferToSend = PooledByteBufAllocator.DEFAULT.compositBuffer() in the beginning, 
then allocate temp buffer for headers and write blob to bufferToSend,
finally return bufferToSend?

Reply:
sure